Feature: Choose a sensible window size on a fresh OTTD config file.

michicc opened the pull request:
## Motivation / Problem

Most screens today are bigger than 640x480. Assume the user actually
wants to see OpenTTD on screen and try to make the window sized
to 75% of the screen.


## Description

If the config file contains a resolution of 0x0 (which is also set as the default value), the video drivers will (where possible) set the actual resolution to 75% of the screen size.

The selected window size will be written to the config after the first start,
disabling this automatic on subsequent launches.

Each video driver must unfortunately duplicate the call to auto-resolution handling, as for example SDL functions can only be called after SDL_Init, which is only done on video driver load when we know SDL is going to be used.

If a video driver does not provide a screen resolution, the old default of 640x480 will be used.


## Limitations

Missing implementation for SDL1, as I couldn't see a proper function for it.


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* ~The bug fix is important enough to be backported? (label: 'backport requested')~
* ~This PR affects the save game format? (label 'savegame upgrade')~
* ~This PR affects the GS/AI API? (label 'needs review: Script API')~
* ~This PR affects the NewGRF API? (label 'needs review: NewGRF')~

michicc committed:
Feature: Choose a sensible window size on a fresh OTTD config file.

Most screens today are bigger than 640x480. Assume the user actually
wants to see OpenTTD on screen and try to make the window sized
to 75% of the screen.

The selected window size will be written to the config after the first start,
disabling this automatic on subsequent launches.

TrueBrain commented:
Why 75%? Most games these days open fullscreen borderless (windowed); we don't have a borderless mode, but why not just the biggest resolution windowed?

TrueBrain commented on src/openttd.cpp:
```diff
@@ -708,14 +708,13 @@ int openttd_main(int argc, char *argv[])
-	_cur_resolution.height = ClampU(_cur_resolution.height, 1, UINT16_MAX / 2);
+	/* Limit resolution as width times height times bytes per pixel must still
+	 * fit within a 32 bits integer, even for 32 bpp video modes. This way all
+	 * internal drawing routines work correctly.
```
Difficult to read sentence. Suggestion:

`Limit the width x height x bytes per pixel to fit in a 32bit integer. This way all internal drawing routines work correctly.`

TrueBrain commented on src/video/video_driver.hpp:
```diff
@@ -102,11 +108,28 @@ class VideoDriver : public Driver {
+	/*
+	 * Get the resolution of the main screen.
+	 */
+	virtual Dimension GetScreenSize() const { return { 640, 480 }; }
```
Shall we bump this res to 1024x768 or something while at it?

michicc commented:
Hysterical raisins? OpenTTD has always defaulted to start windowed (where windows are a thing) and I'm keeping with that. Also, biggest windowed generally means having to use the OS maximize window function, making the whole setting thing more complicated as you need a valid resolution early on.

TrueBrain commented on src/video/video_driver.hpp:
```diff
@@ -102,11 +108,28 @@ class VideoDriver : public Driver {
+	void UpdateAutoResolution()
+	{
+		if (_cur_resolution.width == 0 || _cur_resolution.height == 0) {
+			/* Auto-detect a good resolution. We aim for 75% of the screen size.*
```
dangling star :D

TrueBrain commented on src/video/video_driver.hpp:
```diff
@@ -102,11 +108,28 @@ class VideoDriver : public Driver {
+			/* Auto-detect a good resolution. We aim for 75% of the screen size.*
+			 * Limit resolution as width times height times bytes per pixel must still
+			 * fit within a 32 bits integer, even for 32 bpp video modes. This way all
+			 * internal drawing routines work correctly. */
```
same as earlier

TrueBrain commented on src/video/video_driver.hpp:
```diff
@@ -102,11 +108,28 @@ class VideoDriver : public Driver {
+			 * internal drawing routines work correctly. */
+			Dimension res = this->GetScreenSize();
+			_cur_resolution.width  = ClampU(res.width  * 3 / 4, 640, UINT16_MAX / 2);
+			_cur_resolution.height = ClampU(res.height * 3 / 4, 480, UINT16_MAX / 2);
```
here too, 1024x768? 1280?

TrueBrain reviewed (commented):
I like it :D We can debate settings all day, but in the end, I like this :D So see those remarks as a +0, not a -1 :)

deltanedas commented:
i think 640x480 should be made into two macros at the top so its easier to change without sifting through everything

michicc commented:
I've addressed the comments (pun intended). The default resolution is a constant now, but I didn't change it so far.

It is only used as a lower bound and as a fallback for those video drivers that don't have resolution detection. Right now this is Allegro and SDL1 (and of course null/dedicated), where I don't know on which kind of devices they are still used.
